### ISSUE TITLE
[rewrite] FFmpegPCMAudio's pipe mode fails for internet-originating streams

### ISSUE THREAD

**Issue by saucecode** [Rapptz/discord.py#1065]
**tl;dr** `FFmpegPCMAudio`'s pipe mode will not play audio from an internet source because `subprocess.Popen` mangles the data

I've spent almost a whole day trying to figure out the exact cause of this problem. I'm opening this issue so that others know that this limitation exists and is not caused by discord.py. The problem first occurred when I tried to write/run code that looked a little bit like this.

```python
voice = some_voice_channel.connect() # connect the bot to a voice channel

...

import urllib3
http = urllib3.PoolManager()

# open a stream request to a foreign audio resource
response = http.request('GET', 'http://example.com/audioresource.wav', preload_content=False)
# use pipe mode to direct the bytestream directly to the ffmpeg process
voice.play( discord.FFmpegPCMAudio(response, pipe=True), after=response.close() )
```

The goal here being that you can open a HTTP connection using urllib3 and pipe the download stream directly into the underlying ffmpeg process, and *should* be effectively equivalent to this sample, using a file handle instead of a network stream:

```python
voice = some_voice_channel.connect()

...

audiofile = open('audioresource.wav', 'rb')
voice.play( discord.FFmpegPCMAudio(audiofile, pipe=True), after=audiofile.close )
```
which I can confirm works perfectly.

From what I have determined so far, `subprocess.Popen` outputs mangled data when an stdin pipe's source has a non-trivial latency (such as the latency associated with an online file transfer). As further proof to this, the HTTP stream example does execute correctly if the resource is hosted on a local HTTP server (say, with `python -m SimpleHTTPServer`, and passing in `http://192.168.1.1/audioresource.wav`).

I think this should be mentioned in the documentation, since it takes quite a lot of digging to determine that this limitation is **not** caused by discord.py.

--------

I did develop a workaround for this problem, however it is fairly... it's a hack, pretty much.

```python
voice = some_voice_channel.connect() # connect the bot to a voice channel

...

import urllib3, subprocess
http = urllib3.PoolManager()

# open a stream request to a foreign audio resource
response = http.request('GET', 'http://example.com/audioresource.wav', preload_content=False)
player = discord.FFmpegPCMAudio(subprocess.PIPE, pipe=True)

def download_redirect(resp, process):
    while 1:
        chunk = resp.read(8192)
        if not chunk:
            resp.close()
            process.stdin.close() # tells ffmpeg that there's no more input, triggering voice to call the 'after' callback
            break
        process.stdin.write(chunk)

download_thread = threading.Thread(target=download_redirect, args=(response, player._process))

voice.play( player, after=response.close )
download_thread.start()
```

So we create the FFmpegPCMAudio player and give it a subprocess.PIPE, which eventually becomes the stdin argument in a `Popen` call ([documentation](https://docs.python.org/3/library/subprocess.html#subprocess.Popen)). Then I open a thread to read data from the response and write it directly to the ffmpeg process' stdin pipe.

**Comment by imayhaveborkedit** [comment]
Why not just give ffmpeg the url and let it handle the streaming?  This is what it did for the youtube_dl player in async.

**Comment by saucecode** [comment]
Oh.


God dammit.
